File: heapqueue.py
import copy
import sys
from heapsort import *


class HeapUnderflow(Exception):
    pass


class HeapNewkeyError(Exception):
    pass


class MaxHeapQueue():
    def __init__(self, queue=None):
        if queue is None:
            self.queue = []
        else:
            self.queue = copy.deepcopy(queue)

    def maximum(self):
        return self.queue[0]

    def extract_max(self):
        if len(self.queue) == 0:
            raise HeapUnderflow()
        max = self.queue[0]
        if len(self.queue) > 1:
            self.queue[0] = self.queue.pop()
            max_heapify(self.queue, 0)
        else:
            self.queue.clear()
        return max

    def increase_key(self, i, key):
        if key < self.queue[i]:
            raise HeapNewkeyError()
        self.queue[i] = key
        while(i != 0 and self.queue[parent(i)] < self.queue[i]):
            pi = parent(i)
            self.queue[i], self.queue[pi] = self.queue[pi], self.queue[i]
            i = pi

    def insert(self, key):
        self.queue.append(-sys.maxsize)
        self.increase_key(len(self.queue)-1, key)

    def delete(self, i):
        if len(self.queue) > 1:
            self.queue[i] = self.queue.pop()
            max_heapify(self.queue, i)
        else:
            self.queue.clear()
            

class MinHeapQueue():
    def __init__(self, queue=None):
        if queue is None:
            self.queue = []
        else:
            self.queue = copy.deepcopy(queue)

    def minimum(self):
        return self.queue[0]

    def extract_min(self):
        if len(self.queue) == 0:
            raise HeapUnderflow()
        min = self.queue[0]
        if len(self.queue) > 1:
            self.queue[0] = self.queue.pop()
            min_heapify(self.queue, 0)
        else:
            self.queue.clear()
        return min

    def decrease_key(self, i, key):
        if key > self.queue[i]:
            raise HeapNewkeyError()
        self.queue[i] = key
        while(i != 0 and self.queue[parent(i)] > self.queue[i]):
            pi = parent(i)
            self.queue[i], self.queue[pi] = self.queue[pi], self.queue[i]
            i = pi

    def insert(self, key):
        self.queue.append(sys.maxsize)
        self.decrease_key(len(self.queue)-1, key)

class Item():
    def __init__(self, index, item):
        self.index = index
        self.item = item

    def __int__(self):
        return self.index
        
    def __lt__(self, other):
        return int(self) < int(other)

    def __gt__(self, other):
        return int(self) > int(other)

    def __repr__(self):
        return "{}:{}".format(self.index, self.item)
    
        
class Fifo():
    def __init__(self):
        self.queue = MaxHeapQueue()
        self.index = 0
        
    def size(self):
        return len(self.queue)
        
    def put(self, item):
        self.queue.insert(Item(self.index, item))
        self.index -= 1

    def get(self):
        return self.queue.extract_max().item


def build_max_heap_insert(array):
    heap = MaxHeapQueue()
    for item in array:
        heap.insert(item)
    return heap

File: heapqueue.cpp
#include <iostream>
#include <cmath>
#include <climits>
#include "heapqueue.h"

int parent(int i)
{
    return int((i+1)/2-1);
}

int left(int i){
    return (i+1)*2-1;
}

int right(int i){
    return (i+1)*2;
}


void PrintHeap(const vector<int>& queue){
    int i;
    for(i=0; i<queue.size()-1; ++i)
	cout << queue[i] << ",";
    cout << queue[i] << endl;
}

MaxHeapQueue::MaxHeapQueue(const vector<int>& q)
{
    queue = q;
}

int MaxHeapQueue::Maximum(){
    if(queue.size() == 0)
	throw UnderflowError();
    return queue[0];
}
 
void MaxHeapQueue::Heapify(int i){
    int l = left(i);
    int r = right(i);
    int largest;
    if(l < queue.size() and queue[l] > queue[i]){
	largest = l;
    } else {
	largest = i;
    }
    if(r < queue.size() and queue[r] > queue[largest]){
	largest = r;
    }
    if(largest != i){
	swap(queue[i], queue[largest]);
	Heapify(largest);
    }
}

vector<int> MaxHeapQueue::Queue(){
    return queue;
}

void MaxHeapQueue::BuildHeap(){
    for(int i=int(queue.size()/2)-1; i>=0; --i)
	Heapify(i);
}

vector<int> MaxHeapQueue::HeapSort(){
    vector<int> sorted;
    int size = queue.size();
    sorted.resize(size);
    for(int i=0; i<size; ++i){
	sorted[size-i-1] = queue[0];
	queue[0] = queue[queue.size()-1];
	queue.pop_back();
	Heapify(0);
    }
    return sorted;
}

int MaxHeapQueue::ExtractMax(){
    if(queue.size()==0)
	throw UnderflowError();
    int max = queue[0];
    queue[0] = queue.back();
    queue.pop_back();
    Heapify(0);
    return max;
}

void MaxHeapQueue::IncreaseKey(int i, int key){
    if(key<queue[i])
	throw KeyvalueError();
    queue[i] = key;
    while(i>0 and queue[parent(i)] < queue[i]){
	swap(queue[parent(i)], queue[i]);
	i = parent(i);
    }
}

void MaxHeapQueue::Insert(int key){
    queue.push_back(INT_MIN);
    IncreaseKey(queue.size()-1, key);
}

